Updating password rules

muralibasani opened the pull request:
# Linked issue

Resolves: #xxxxx

# What kind of change does this PR introduce?

- [ ] Bug fix
- [ ] New feature
- [ ] Refactor
- [ ] Docs update
- [ ] CI update

# What is the current behavior?

_Describe the state of the application before this PR. Illustrations appreciated (videos, gifs, screenshots)._

# What is the new behavior?

_Describe the state of the application after this PR. Illustrations appreciated (videos, gifs, screenshots)._

# Other information

_Additional changes, explanations of the approach taken, unresolved issues, necessary follow ups, etc._

# Requirements (all must be checked before review)

- [ ] The pull request title follows [our guidelines](https://github.com/Aiven-Open/klaw/blob/main/CONTRIBUTING.md#guideline-commit-messages)
- [ ] Tests for the changes have been added (if relevant)
- [ ] The latest changes from the `main` branch have been pulled
- [ ] `pnpm lint` has been run successfully

ArtturiSipila commented on core/src/main/resources/templates/addUser.html:
```diff
@@ -504,7 +504,12 @@ <h4 class="mb-3">Shortcuts</h4>
+													   minlength="8"
+													   required ng-model="addNewUser.pwd"
+													   pattern="(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[\W_]).{8,}"
+													   title="Password must be at least 8 characters long and include at least one uppercase letter, one lowercase letter, one number, and one special character."
```
Is it enough to check it at the frontend?

muralibasani replied:
It is better to add in backend as well. Let me do that with a couple of tests.

ArtturiSipila replied:
Great, thanks!

nosahama commented on coral/src/app/features/user-information/change-password/changePasswordFormSchema.ts:
```diff
@@ -4,7 +4,13 @@ const changePasswordFormSchema = z
+      .regex(/[a-z]/, { message: "Must include at least one lowercase letter" })
+      .regex(/\d/, { message: "Must include at least one number" })
+      .regex(/[\W_]/, {
+        message: "Must include at least one special character",
```
**[nit]**
why the new line 😄

muralibasani replied:
that's coming with linting/format :)

nosahama commented on coral/types/api.d.ts:
```diff
@@ -2709,7 +2709,7 @@ export type components = {
             username: string;
             role: string;
-            userPassword: string;
+            userPassword?: string;
```
**[inquiry]**
`userPassword` is not required?

nosahama commented on coral/types/api.d.ts:
```diff
@@ -3091,8 +3091,8 @@ export type components = {
         ChangePasswordRequestModel: {
-            pwd: string;
-            repeatPwd: string;
+            pwd?: string;
```
[inquiry]
for a password change model, the passwords should be required, even if the initial password can be blank for other technical reasons, password reset must always receive a value, wdyt?

muralibasani replied:
It should not be the case. Fixed.

nosahama commented on core/src/main/java/io/aiven/klaw/model/requests/ChangePasswordRequestModel.java:
```diff
@@ -12,11 +13,15 @@
+  @Pattern(
+      regexp = PASSWORD_REGEX,
+      message =
+          "Password must be at least 8 characters long and include at least one uppercase letter, one lowercase letter, one number, and one special character.")
```
**[extract]**
This text is repeated here, `core/src/main/java/io/aiven/klaw/model/requests/RegisterUserInfoModel.java` and `core/src/main/java/io/aiven/klaw/model/requests/UserInfoModel.java`, maybe extract it into it's own object just like the `PASSWORD_REGEX`.

nosahama replied:
We could even have a custom decorator `PasswordPattern` which will abstract away this details.

muralibasani replied:
For now, just created a constant.

nosahama commented on core/src/main/resources/static/js/manageUsers.js:
```diff
@@ -114,6 +114,12 @@ app.controller("manageUsersCtrl", function($scope, $http, $location, $window) {
             serviceInput['pwd'] = $scope.chPwd.pwd;
             serviceInput['repeatPwd'] = $scope.chPwd.repeatpwd;
 
+            if(!$scope.chPwd.pwd) {
```
**[inquiry]**
could this check be done before the `serviceInput` object is populated?

muralibasani replied:
Updated.

nosahama commented on core/src/main/resources/templates/addUser.html:
```diff
@@ -504,7 +504,12 @@ <h4 class="mb-3">Shortcuts</h4>
 											<div class="form-group">
 												<label>Password</label>
-												<input type="password" class="form-control" minlength="8" required ng-model="addNewUser.pwd">
+												<input type="password" class="form-control"
```
**[nit]**
Some formatting will be nice 😄

nosahama commented on core/src/test/java/io/aiven/klaw/controller/UiConfigControllerTest.java:
```diff
@@ -404,4 +403,24 @@ public void showActivityLog() throws Exception {
+
+  @Test
+  @Order(21)
+  public void changePwdSuccess() throws Exception {
```
**[inquiry]**
I am unsure where the `Exception` is thrown, also in `changePwdFailure` above, since we are checking the return status.

muralibasani replied:
Its the object mapper and mvc perform methods which could throw.

nosahama replied:
Thank you.

nosahama commented on core/src/test/java/io/aiven/klaw/service/UsersTeamsControllerServiceTest.java:
```diff
@@ -1315,7 +1340,7 @@ public void registerUserInternal(
     model.setTeam("Octopus");
     model.setRole("USER");
-    model.setPwd("XXXXXXX");
+    model.setPwd("XXXXXXX12@");
```
**[inquiry]**
I am guessing this is a failiing test as there is no lowercase letter in this password.

nosahama replied:
`Password must be at least 8 characters long and include at least one uppercase letter, one lowercase letter, one number, and one special character.` 🤔

muralibasani replied:
Missed it. You are right.
This should be good now. As the test is testing an internal method, it did not throw any expception.
But apis/controllers are in good shape.

nosahama commented on openapi.yaml:
```diff
@@ -6517,7 +6517,7 @@
           }
         },
-        "required" : [ "fullname", "username" ]
+        "required" : [ "fullname", "pwd", "username" ]
```
**[inquiry]**
isn't `pwd` supposed to be `userPassword`? And required as per this discussion - https://github.com/Aiven-Open/klaw/pull/2842#discussion_r2000616949?

muralibasani replied:
So this is a generated file. And, it is part of RegisterUserInfoModel. Should be ok.
Before the last commit, it was optional, but fixed it back to mandatory.

aindriu-aiven commented on core/src/main/java/io/aiven/klaw/model/requests/ChangePasswordRequestModel.java:
```diff
@@ -12,11 +15,11 @@
 
-  @Size(min = 8)
   @NotNull
+  @Pattern(regexp = PASSWORD_REGEX, message = PASSWORD_REGEX_VALIDATION_STR)
```
It might be worth splitting the @Pattern into 3 or 4 things to give concise feedback about the issue with the password wdyt?

e.g. 
@Pattern(regex =  "(?=.*[a-z])" message = "must contain at least 1 lower case letter")
@Pattern(regex =  "(?=.*[A-Z])" message = "must contain at least 1 upper case letter")
@Pattern(regex =  "(?=.*[0-9])" message = "must contain at least 1 number")
@Pattern(regex =  "(?=.*["!@#$%&*()'+,-./:;<=>?[]^_`{|}"])" message = "must contain at least 1 special character")
@Size(min = 8)


Note have not tested above regex

muralibasani replied:
We have the same pattern check in front end.
This pattern is invoked only when someone is invoking BE apis manually, and we have not enabled them yet.

aindriu-aiven commented on core/src/main/resources/static/js/manageUsers.js:
```diff
@@ -336,7 +342,7 @@ app.controller("manageUsersCtrl", function($scope, $http, $location, $window) {
             if(!$scope.addNewUser.pwd)
             {
-                $scope.alertnote = "Please enter a password.";
+                $scope.alertnote = "Please enter a suggested password.";
```
i dont think you need to add suggested here, as they should be entering their password?

aindriu-aiven commented on core/src/main/resources/static/js/registerUsers.js:
```diff
@@ -389,7 +389,7 @@ app.controller("registerUsersCtrl", function($scope, $http, $location, $window)
           if(!$scope.registerUser.pwd)
           {
-              $scope.alertnote = "Please enter a password.";
+              $scope.alertnote = "Please enter a suggested password.";
```
same I dont think you need to change it to suggested here

aindriu-aiven reviewed (commented):
A few comments for consideration.

aindriu-aiven reviewed (changes requested):
Tested changing password through old and new UI and tested creating a new user, having admin reset password as well.

The one area I fond that wasnt covered by the improved password policy was the password reset page.
Here any password can be added without validation.

UserTeamsControllerService and i think forgotPassword.js need to be updated

aindriu-aiven commented:
> Tested changing password through old and new UI and tested creating a new user, having admin reset password as well.
> 
> The one area I fond that wasnt covered by the improved password policy was the password reset page. Here any password can be added without validation.
> 
> UserTeamsControllerService and i think forgotPassword.js need to be updated

@muralibasani you can just make the change on the API to check it, the error should be propagated back to the UI.

muralibasani commented:
> Tested changing password through old and new UI and tested creating a new user, having admin reset password as well.
> 
> The one area I fond that wasnt covered by the improved password policy was the password reset page. Here any password can be added without validation.
> 
> UserTeamsControllerService and i think forgotPassword.js need to be updated

Right, updated and added a test.
